p2p/transport/webrtc/udpmux/mux.go

```diff
@@ -0,0 +1,309 @@
+// UDP addresses of the same IP address family (eg. Server-reflexive addresses
+// and peer-reflexive addresses).
+func (mux *udpMux) GetConn(ufrag string, addr net.Addr) (net.PacketConn, error) {
+	a, ok := addr.(*net.UDPAddr)
+	if !ok && addr != nil {
+		return nil, fmt.Errorf("unexpected address type: %T", addr)
+	}
+	isIPv6 := ok && a.IP.To4() == nil
+	return mux.getOrCreateConn(ufrag, isIPv6)
+}
+
+// Close implements ice.UDPMux
+func (mux *udpMux) Close() error {
+	select {
+	case <-mux.ctx.Done():
+		return nil
+	default:
+	}
+	mux.cancel()
+	mux.socket.Close()
+	mux.wg.Wait()
+	return nil
+}
+
+// RemoveConnByUfrag implements ice.UDPMux
+func (mux *udpMux) RemoveConnByUfrag(ufrag string) {
+	if ufrag != "" {
+		mux.storage.RemoveConnByUfrag(ufrag)
+	}
+}
+
+func (mux *udpMux) getOrCreateConn(ufrag string, isIPv6 bool) (net.PacketConn, error) {
+	select {
+	case <-mux.ctx.Done():
+		return nil, io.ErrClosedPipe
+	default:
+		return mux.storage.GetOrCreateConn(ufrag, isIPv6, mux)
+	}
+}
+
+// writeTo writes a packet to the underlying net.PacketConn
+func (mux *udpMux) writeTo(buf []byte, addr net.Addr) (int, error) {
+	return mux.socket.WriteTo(buf, addr)
+}
+
+func (mux *udpMux) readLoop() {
+	for {
+		select {
+		case <-mux.ctx.Done():
+			return
+		default:
+		}
+
+		buf := pool.Get(ReceiveMTU)
+
+		n, addr, err := mux.socket.ReadFrom(buf)
+		if err != nil {
+			log.Errorf("error reading from socket: %v", err)
+			pool.Put(buf)
+			if os.IsTimeout(err) {
+				continue
+			}
+			return
+		}
+		buf = buf[:n]
+
+		// a non-nil error signifies that the packet was not
+		// passed on to any connection, and therefore the current
+		// function has ownership of the packet. Otherwise, the
+		// ownership of the packet is passed to a connection
+		processErr := mux.processPacket(buf, addr)
+		if processErr != nil {
+			buf = buf[:cap(buf)]
+			pool.Put(buf)
+		}
+	}
+}
+
+func (mux *udpMux) processPacket(buf []byte, addr net.Addr) error {
+	udpAddr, ok := addr.(*net.UDPAddr)
+	if !ok {
+		return fmt.Errorf("underlying connection did not return a UDP address")
+	}
+	isIPv6 := udpAddr.IP.To4() == nil
+
+	// Connections are indexed by remote address. We firest
+	// check if the remote address has a connection associated
+	// with it. If yes, we push the received packet to the connection
+	// and loop again.
+	conn, ok := mux.storage.GetConnByAddr(udpAddr)
+	// if address was not found check if ufrag exists
+	if !ok && stun.IsMessage(buf) {
+		msg := &stun.Message{Raw: buf}
+		if err := msg.Decode(); err != nil || msg.Type != stun.BindingRequest {
+			log.Debug("incoming message should be a STUN binding request")
+			return err
+		}
+
+		ufrag, err := ufragFromStunMessage(msg)
+		if err != nil {
+			log.Debug("could not find STUN username: %w", err)
+			return err
+		}
+
+		var connCreated bool
+		conn, connCreated = mux.storage.AddAddr(ufrag, udpAddr, isIPv6, mux)
+
+		if connCreated && mux.unknownUfragCallback != nil {
```

Comment:
`unknownUfragCallback` is always `!= nil`, isn't it?

Reply:
Should this be `!connCreated` here?

Reply:
No it isn't always nil. E.g. in test code it isn't. The typing also doesn't guarantee it's non-nil.
And no it's unknown if it wasn't created before, as such connCreated.